src/Propel/Generator/Behavior/Archivable/ArchivableBehavior.php

```diff
@@ -90,79 +92,176 @@ public function modifyTable(): void
         }
     }
 
+    /**
+     * @return string
+     */
+    protected function getArchiveTableName(): string
+    {
+        return $this->getParameter('archive_table') ?: ($this->getTable()->getOriginCommonName() . '_archive');
+    }
+
     /**
      * @return void
      */
     protected function addArchiveTable(): void
     {
         $table = $this->getTable();
         $database = $table->getDatabase();
+        $archiveTableName = $this->getArchiveTableName();
+
+        $archiveTableExistsInSchema = $database->hasTable($archiveTableName);
+
+        $this->archiveTable = $archiveTableExistsInSchema ?
+            $database->getTable($archiveTableName) :
+            $this->createArchiveTable();
+
+        if ($archiveTableExistsInSchema && !$this->parameterHasValue('sync', 'true')) {
+            return;
+        }
+
+        $this->syncTables();
+    }
+
+    /**
+     * @return \Propel\Generator\Model\Table
+     */
+    protected function createArchiveTable(): Table
+    {
+        $sourceTable = $this->getTable();
+        $database = $sourceTable->getDatabase();
+
+        // create the version table
+        return $database->addTable([
+            'name' => $this->getArchiveTableName(),
+            'phpName' => $this->getParameter('archive_phpname'),
+            'package' => $sourceTable->getPackage(),
+            'schema' => $sourceTable->getSchema(),
+            'namespace' => $sourceTable->getNamespace() ? '\\' . $sourceTable->getNamespace() : null,
+            'identifierQuoting' => $sourceTable->isIdentifierQuotingEnabled(),
+        ]);
+    }
+
+    /**
+     * @return \Propel\Generator\Model\Table
+     */
+    protected function syncTables(): Table
+    {
+        $archiveTable = $this->getArchiveTable();
+        $sourceTable = $this->getTable();
+        $database = $sourceTable->getDatabase();
         $platform = $database->getPlatform();
-        $archiveTableName = $this->getParameter('archive_table') ?: ($this->getTable()->getOriginCommonName() . '_archive');
-        if (!$database->hasTable($archiveTableName)) {
-            // create the version table
-            $archiveTable = $database->addTable([
-                'name' => $archiveTableName,
-                'phpName' => $this->getParameter('archive_phpname'),
-                'package' => $table->getPackage(),
-                'schema' => $table->getSchema(),
-                'namespace' => $table->getNamespace() ? '\\' . $table->getNamespace() : null,
-                'identifierQuoting' => $table->isIdentifierQuotingEnabled(),
+
+        // copy all the columns
```

Comment:
Good job so far!

Can we please move all exact functionality for their own methods to meet SRP?
So in the end we have this method something like
```
$this->syncColumns($sourceTable, $archiveTable);
$this->syncForeignKeys($sourceTable, $archiveTable);
$this->syncIndixes($sourceTable, $archiveTable);
// so on
```

Thank you in advance!

Reply:
Thank you for the review!
Method is split up now.